### ISSUE TITLE
Add injection compressor example

### ISSUE THREAD

**Issue by michmest** [oemof/tespy#230]
Hello 

I'm doing my master thesis about vapor injection in compressors. To make a model of the setup I'm using your library for which I am extremely grateful. But unfortunately, the simulation does not converge. The system layout is based on a hot gas bypass stand with 2 extra injection lines that mix right before injection into the compressor.  As in this paper [design_testbench.pdf](https://github.com/oemof/tespy/files/5435898/design_testbench.pdf). Where the compression with injection is modeled as a sequence of a compressor, a merging and another compressor. 

I read in the docs that initialization is very important to convergence. So first I simulated the hot gas bypass stand without injection. This does work but has some high zeta values due to low mass flow rates I assume. Can the max value of zeta be changed somewhere? Then I use this result to initialize the hot gas bypass stand with injection but the solution doesn't converge. With the error message: "WARNING-The solver does not seem to make any progress, aborting calculation. Residual value is 1.08e+05. This frequently happens, if the solver pushes the fluid properties out of their feasible range." Do you have any ideas I could try to keep the fluid properties within a feasible range?

Here is my work so far [HotGasBypassWithEconomization.zip](https://github.com/oemof/tespy/files/5435899/HotGasBypassWithEconomization.zip). 

Kind regards

Michiel

**Comment by fwitte** [comment]
Hi @michmest,
which version of TESPy are you using? With the latest version, I can not replicate your issue, as the small cycle does not even converge. However, with some modifications, I got the small cycle running.

Generally, rather than building the full cycle at once you should build it up step by step making small additions and rerun the simulation. By this, you can identify possible sources of instability.
Best
Francesco

**Comment by fwitte** [comment]
Ok, I got it working with every version since 0.3.0 now. It seems like the power specification was one of the major issues. I needed three steps in total.

Good luck with your thesis and have fun working on TESPy :). Appreciate your feedback!

Here is my code, I added some plots:

```python
# -*- coding: utf-8 -*-
"""
Created on Sun Oct 11 19:09:07 2020

@author: Michiel
"""
from tespy.networks import network
from tespy.components import (
    sink, source, compressor, heat_exchanger_simple,
    valve, cycle_closer, splitter, merge
)
from tespy.tools import logger
import logging
from tespy.tools import dc_prop
from tespy.connections import connection, ref, bus
import CoolProp.CoolProp as CP

from tespy.tools.characteristics import char_line
import numpy as np
import pandas as pd

try:
    from fluprodia import FluidPropertyDiagram
    plotting = True
except ModuleNotFoundError:
    plotting = False

# %% logger

logger.define_logging(
    logfile='tespy.log',
    log_path=True, log_version=True,
    screen_level=logging.INFO, file_level=logging.DEBUG
)

# %% Input parameters model

Refrigerant = 'R32'
T_cd = 273.15 + 20 # K condenser temperature
T_ev = 273.15 - 15 # K evaporater temperature
dT_SH = 10 # K superheat compressor inlet
dT_SC = -5 # K subcooling condenser outlet
P_co = 1200 # W compresser power input
Q_c = 2e3 # W cooling capacity
eta_is = 0.7 # isentropic efficiency
# eta_v = 1 # volumetric efficiency
# V_d = # cc compressor volumetric displacement
# n =  # rpm compressor speed
Q_d = 4 # m^3/h volumetric flow rate
R_inj = 0.05 # injection mass flow rate/ suction mass flow rate
p_inj = (CP.PropsSI('P', 'T', T_cd, 'Q', 1, 'R32')*CP.PropsSI('P', 'T', T_ev, 'Q', 1, 'R32')) ** 0.5 # injection pressure

# %% networks

# hot gas bypass stand
hgbs = network(
    fluids=['R32'], T_unit='K', p_unit='Pa', h_unit='kJ / kg',
    m_unit='kg / s'
)

# %% components

# sources & sinks
cc = cycle_closer('coolant cycle closer')
# cc = source('dunmy source')

# splitters and mergers
sp1 = splitter('splitter 1', num_out = 2)
sp2 = splitter('splitter 2', num_out = 2)
sp3 = splitter('splitter 3', num_out = 2)

me1 = merge('merger 1', num_in = 2)
me2 = merge('merger 2', num_in = 2)
me3 = merge('merger 3', num_in = 2)

# condenser
cd = heat_exchanger_simple('condenser')

# valves
val = []
for i in range(5):
    val.append(valve('EEV' + str(i)))

# compressor-system
cp1 = compressor('compressor 1')
cp2 = compressor('compressor 2')

# %% connections

# shared connections
cc_val0 = connection(cc, 'out1', val[0], 'in1')
val0_sp1 = connection(val[0], 'out1', sp1, 'in1')

sp1_val1 = connection(sp1, 'out1', val[1], 'in1')
sp1_cd = connection(sp1, 'out2', cd, 'in1')
cd_val2 = connection(cd, 'out1', val[2], 'in1')
me1_s = connection(me1, 'out1', sink('test1'), 'in1')

val1_me1 = connection(val[1], 'out1', me1, 'in1')
val2_me1 = connection(val[2], 'out1', me1, 'in2')
me1_cp1 = connection(me1, 'out1', cp1, 'in1')

cp1_cp2 = connection(cp1, 'out1', cp2, 'in1')
cp2_s = connection(cp2, 'out1', sink('test'), 'in1')
cp2_cc = connection(cp2, 'out1', cc, 'in1')

hgbs.add_conns(cc_val0,val0_sp1,sp1_val1, sp1_cd, cd_val2, val2_me1, val1_me1, me1_cp1, cp1_cp2, cp2_cc)

# %% busses

cp_bus = bus('compressor')
cp_bus.add_comps({'comp': cp1}, {'comp': cp2})
hgbs.add_busses(cp_bus)

# %% component parametrization

# condenser
cd.set_attr(pr=1)

# compressor system
cp1.set_attr(eta_s = eta_is)
cp2.set_attr(eta_s = eta_is)

# %% connection parametrization

# fluid
cc_val0.set_attr(fluid={'R32' : 1}, p=15e5)
val0_sp1.set_attr()
# specified mass flow yields high stability
sp1_val1.set_attr(m=0.02)

# condenser pressure and condenser subcooling
cd_val2.set_attr(p = CP.PropsSI('P', 'T', T_cd, 'Q', 1, 'R32'), Td_bp=dT_SC)

# compressor 1 input pressure and volumetric flow
me1_cp1.set_attr(p=CP.PropsSI('P', 'T', T_ev, 'Q', 0, 'R32'), v=Q_d/3600)
cp1_cp2.set_attr(p = p_inj)

# %% Calculation

hgbs.solve('design')

# replace the initial mass flow with temperature difference to boiling point
# note: h0 required even though we have good starting values. This should
# not be the case. Is fixed in version 0.3.5.dev0
me1_cp1.set_attr(Td_bp=dT_SH, h0=600)
sp1_val1.set_attr(m=None)

# add missing parts of the cycle
hgbs.del_conns(sp1_cd, cd_val2)
sp1_sp2 = connection(sp1, 'out2', sp2, 'in1')
sp2_val3 = connection(sp2, 'out1', val[3], 'in1')
sp2_cd = connection(sp2, 'out2', cd, 'in1')

val3_me2 = connection(val[3], 'out1', me2, 'in1')

cd_sp3 = connection(cd, 'out1', sp3, 'in1')
sp3_val2 = connection(sp3, 'out1', val[2], 'in1')
sp3_val4 = connection(sp3, 'out2', val[4], 'in1')

val4_me2 = connection(val[4], 'out1', me2, 'in2')
me2_me3 = connection(me2, 'out1', sink('dummy me3'), 'in1')

# use a dummy for the merge 3 for stability
hgbs.add_conns(sp1_sp2, sp2_cd, cd_sp3, sp3_val2, sp3_val4, val4_me2, sp2_val3, val3_me2, me2_me3)

# re-set condenser outlet state
cd_sp3.set_attr(p=CP.PropsSI('P', 'T', T_cd, 'Q', 1, 'R32'), Td_bp=dT_SC)

# some mass flow must be set here, as it is a loose branch
sp2_val3.set_attr(m=1)

me2_me3.set_attr(p=p_inj, x=0.5)


hgbs.solve('design')

# finish cycle
hgbs.del_conns(cp1_cp2, me2_me3)
cp1_me3 = connection(cp1, 'out1', me3, 'in1')
me3_cp2 = connection(me3, 'out1', cp2, 'in1')
me2_me3 = connection(me2, 'out1', me3, 'in2')
hgbs.add_conns(cp1_me3, me3_cp2, me2_me3)

sp2_val3.set_attr(m=None) # unset the mass flow again

# respecify merge2 to merge3 attributes as old connection was deleted
me2_me3.set_attr(p=p_inj, x=0.5)
# only missing: some inlet state for compressor 2
# note: h0 required even though we have good starting values. This should
# not be the case. Is fixed in version 0.3.5.dev0
me3_cp2.set_attr(Td_bp=10, h0=600)
hgbs.solve('design')

# Specifying the compressor power input lastly
# The pressure specification at the outlet of the compressor 2 is
# now performed by the specified total power input. This is one of the main
# sources of instability
cp_bus.set_attr(P=P_co)
cc_val0.set_attr(p=None)

hgbs.solve('design')
hgbs.print_results()

# plot the results into logph and Ts diagrams
result_dict = {
    prop: [
        conn.get_attr(prop).val for conn in hgbs.conns.index
    ] for prop in ['p', 'h', 's', 'T']
}

if plotting is True:
    diagram = FluidPropertyDiagram('R32')
    diagram.set_unit_system(T='°C', p='bar', h='kJ/kg')
    iso_T = np.arange(-50, 81, 10)
    diagram.set_isolines(T=iso_T)
    diagram.calc_isolines()
    diagram.calc_isolines()
    diagram.set_limits(x_min=500, x_max=2750, y_min=-40, y_max=100)
    diagram.draw_isolines('Ts')
    diagram.ax.scatter(result_dict['s'], np.array(result_dict['T']) - 273.15)
    diagram.save('R32_Ts.pdf')

    diagram.set_limits(x_min=100, x_max=650, y_min=1, y_max=100)
    diagram.draw_isolines('logph')
    diagram.ax.scatter(result_dict['h'], np.array(result_dict['p']) / 1e5)
    diagram.save('R32_logph.pdf')

```
![R32_logph](https://user-images.githubusercontent.com/20252362/97169152-dc5d6a00-1789-11eb-9ae4-55561e396d9b.png)
![R32_Ts](https://user-images.githubusercontent.com/20252362/97169155-dd8e9700-1789-11eb-9029-7a285daf55a2.png)

**Comment by fwitte** [comment]
P. S.: Just thinking: As the cycle seems quite fiddly to set up, this would make a good example or tutorial. If you are interested in sharing results form your thesis, this might be a nice addition to the online-documentation :).
Best regards, have a nice week

**Comment by michmest** [comment]
Thank you so much, that's amazing! I couldn't figure out what the problem was. Honestly, this library together with fluprodia is so useful. Great work. I had one more question about plotting states in fluprodia. Is there a way to select between which states a line must be drawn? Since the system isn't a single cycle, the result I get using the plot() function is a bit weird.
About the results from my thesis. I will first have to ask my supervisors but I assume this won't be a problem. Thank you once again.

Kind regards

Michiel

**Comment by fwitte** [comment]
Hi Michiel,
good to hear that, thank you!

So my issue for drawing specific lines in fluprodia does not lie in drawing isenthalp, isobar, isochor, isotherm or isentropic lines, but in drawing lines between states that do not connect with a "simple" process. For example, consider a condenser with pressure losses on the hot side and overheated inlet state: What share of the pressure losses is seen until saturated gaseous state and what share in the actual condensing? How would a line be drawn between two states that connect via non-isentropic (isentropic efficiency < 1) expansion or compression? Using polytropic lines could be the answer, but unfortunately I do not have the time to investigate this at the moment.

In respect to your issue: the .plot() function of matplotlib.pyplot does automatically draw straight lines between the points you pass to the function in the order you pass the points. Therefore you would need to pass each inlet and outlet state of a component individually. But: The points do not necessarily connect with straight lines! As your example consists mainly consists isobars and isenthalps, I would actually suggest (and prefer of course), that we update the fluprodia package to be able to draw individual simple processes. If you are interested in contributing, I opened an issue (https://github.com/fwitte/fluprodia/issues/3), where we can continue to discuss this topic.

Best regards
Francesco

P. S.: Regarding you initial question on changing the value limits: `yourcomponent.zeta.set_attr(max_val=...)`

**Comment by fwitte** [comment]
Hi,
I added the functionailty for plotting individual isolines to fluprodia. You can check it out at https://fluprodia.readthedocs.io/en/latest/usage.html, there is a simple example on TESPy-FluProDia integration, too.
Best

**Comment by fwitte** [comment]
With the next TESPy version release an automatic processing of the input data required by FluProDia to generate the connecting line data will be implemented (see #234). The code could look like this:

``` python
# -*- coding: utf-8 -*-

[...]
Your TESPy script to calculate the plant
[...]

tespy_results = {
    c.label: c.get_plotting_data() for c in hgbs.components.values()
    if len(c.get_plotting_data()) > 0
}


diagram = FluidPropertyDiagram('R32')
diagram.set_unit_system(T='K', p='Pa', h='kJ/kg')
iso_T = np.arange(250, 451, 25)
diagram.set_isolines(T=iso_T)
diagram.calc_isolines()


for key, data in tespy_results.items():
    try:
        tespy_results[key]['datapoints'] = diagram.calc_individual_isoline(**data)
    except TypeError:
        for nested_key, nested_data in tespy_results[key].items():
            tespy_results[key][nested_key]['datapoints'] = diagram.calc_individual_isoline(**nested_data)


diagram.set_limits(x_min=100, x_max=750, y_min=4e5, y_max=3e6)
diagram.draw_isolines('logph')

for key in tespy_results.keys():
    if 'datapoints' in tespy_results[key].keys():
        datapoints = tespy_results[key]['datapoints']
        diagram.ax.plot(datapoints['h'], datapoints['p'], color='#ff0000')
        diagram.ax.scatter(datapoints['h'][0], datapoints['p'][0], color='#ff0000')
    else:
        for nested_key in tespy_results[key].keys():
            datapoints = tespy_results[key][nested_key]['datapoints']
            diagram.ax.plot(datapoints['h'], datapoints['p'], color='#ff0000')
            diagram.ax.scatter(datapoints['h'][0], datapoints['p'][0], color='#ff0000')
diagram.save('logph_diagram_states.svg')

diagram.set_limits(x_min=800, x_max=2500, y_min=250, y_max=450)
diagram.draw_isolines('Ts')

for key in tespy_results.keys():
    if 'datapoints' in tespy_results[key].keys():
        datapoints = tespy_results[key]['datapoints']
        diagram.ax.plot(datapoints['s'], datapoints['T'], color='#ff0000')
        diagram.ax.scatter(datapoints['s'][0], datapoints['T'][0], color='#ff0000')
    else:
        for nested_key in tespy_results[key].keys():
            datapoints = tespy_results[key][nested_key]['datapoints']
            diagram.ax.plot(datapoints['s'], datapoints['T'], color='#ff0000')
            diagram.ax.scatter(datapoints['s'][0], datapoints['T'][0], color='#ff0000')
diagram.save('Ts_diagram_states.svg')
```

![logph_diagram_states](https://user-images.githubusercontent.com/20252362/98519254-c911ca80-2270-11eb-9b8d-0e691e4b2002.png)
![Ts_diagram_states](https://user-images.githubusercontent.com/20252362/98519259-cadb8e00-2270-11eb-9cc2-ce561adecf9e.png)

**Comment by fwitte** [comment]
Hi @michmest,
hope your project did go well. Can we close this issue?
Best regards
